pkg/offsetguess/offsetguess.go

```diff
@@ -0,0 +1,304 @@
+// limitations under the License.
+
+// +build linux
+
+package offsetguess
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/iovisor/gobpf/elf"
+
+	"github.com/weaveworks/tcptracer-bpf/pkg/byteorder"
+)
+
+type tcpTracerState uint64
+
+const (
+	uninitialized tcpTracerState = iota
+	checking
+	checked
+	ready
+)
+
+type guessWhat uint64
+
+const (
+	guessSaddr guessWhat = iota
+	guessDaddr
+	guessFamily
+	guessSport
+	guessDport
+	guessNetns
+	guessDaddrIPv6
+)
+
+type tcpTracerStatus struct {
+	status          tcpTracerState
+	pidTgid         uint64
+	what            guessWhat
+	offsetSaddr     uint64
+	offsetDaddr     uint64
+	offsetSport     uint64
+	offsetDport     uint64
+	offsetNetns     uint64
+	offsetIno       uint64
+	offsetFamily    uint64
+	offsetDaddrIPv6 uint64
+	err             byte
+	saddr           uint32
+	daddr           uint32
+	sport           uint16
+	dport           uint16
+	netns           uint32
+	family          uint16
+	daddrIPv6       [4]uint32
+}
+
+func listen(url, netType string, finish chan struct{}) {
+	l, err := net.Listen(netType, url)
+	if err != nil {
+		panic(err)
+	}
+	select {
+	case <-finish:
+		l.Close()
+		return
+	}
+}
+
+func compareIPv6(a, b [4]uint32) bool {
+	for i := 0; i < 4; i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func ownNetNS() (uint64, error) {
+	var s syscall.Stat_t
+	if err := syscall.Stat("/proc/self/ns/net", &s); err != nil {
+		return 0, err
+	}
+	return s.Ino, nil
+}
+
+func ipFromUint32Arr(ipv6Addr [4]uint32) net.IP {
+	buf := make([]byte, 16)
+	for i := 0; i < 16; i++ {
+		buf[i] = *(*byte)(unsafe.Pointer((uintptr(unsafe.Pointer(&ipv6Addr[0])) + uintptr(i))))
+	}
+	return net.IP(buf)
+}
+
+func htons(a uint16) uint16 {
+	arr := make([]byte, 2)
+	binary.BigEndian.PutUint16(arr, a)
+	return byteorder.Host.Uint16(arr)
+}
+
+// Guess expects elf.Module to hold a tcptracer-bpf object and initializes
+// the tracer by guessing the right kernel struct offsets. Results are
+// stored in the `tcptracer_status` map as used by the module.
+func Guess(b *elf.Module) error {
+	listenIP := "127.0.0.2"
+	listenPort := uint16(9091)
+	bindAddress := fmt.Sprintf("%s:%d", listenIP, listenPort)
+
+	finish := make(chan struct{})
+
+	currentNetns, err := ownNetNS()
+	if err != nil {
+		return fmt.Errorf("error getting current netns: %v", err)
+	}
+
+	mp := b.Map("tcptracer_status")
+
+	var zero uint64
+	pidTgid := uint64(os.Getpid()<<32 | syscall.Gettid())
+
+	status := tcpTracerStatus{
+		status:  checking,
+		pidTgid: pidTgid,
+	}
+
+	// if we already have the offsets, just return
+	err = b.LookupElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(&status))
+	if err == nil && status.status == ready {
+		return nil
+	}
+
+	go listen(bindAddress, "tcp4", finish)
+	defer close(finish)
+	time.Sleep(300 * time.Millisecond)
```

Comment:
I think it is waiting until the go routine listen() has finished the listen() syscall so that the main program can connect. I will synchronize this with a go channel instead.